openforcefield/tests/test_parameters.py

```diff
@@ -300,6 +466,105 @@ def test_round_trip(self):
+        assert all(isinstance(parameter_attributes[name], IndexedParameterAttribute)
+                   for name in expected_names)
+
+    def test_find_all_required_and_optional_parameter_attrs(self):
+        """ParameterType distinguish between required and optional ParameterAttributes."""
+        class MyParameter(ParameterType):
+            required = ParameterAttribute()
+            optional = ParameterAttribute(default=1)
+            required_indexed = IndexedParameterAttribute()
+            optional_indexed2 = IndexedParameterAttribute(default=None)
+
+        expected_names = ['smirks', 'required', 'required_indexed']
+        parameter_attributes = MyParameter._get_required_parameter_attributes()
+        assert list(parameter_attributes.keys()) == expected_names
+
+        expected_names = ['id', 'parent_id', 'optional', 'optional_indexed2']
+        parameter_attributes = MyParameter._get_optional_parameter_attributes()
+        assert list(parameter_attributes.keys()) == expected_names
+
+    def test_required_attribute_on_init(self):
+        """ParameterType raises TypeError if a required attribute is not specified on construction."""
+        class MyParameter(ParameterType):
+            required = ParameterAttribute()
+            optional = ParameterAttribute(default=None)
+        with pytest.raises(SMIRNOFFSpecError, match="require the following missing parameters"):
+            MyParameter(smirks='[*:1]', optional=1)
+
+    def test_indexed_attrs(self):
+        """ParameterType handles indexed attributes correctly."""
+        class MyParameter(ParameterType):
+            a = IndexedParameterAttribute()
+            b = IndexedParameterAttribute()
+        my_par = MyParameter(smirks='[*:1]', a1=1, a3=3, a2=2, b1=4, b2=5, b3=6)
+        assert my_par.a == (1, 2, 3)
+        assert my_par.b == (4, 5, 6)
+
+    def test_single_value_indexed_attr(self):
+        """ParameterType handle single-value indexed attributes correctly."""
+        class MyParameter(ParameterType):
+            a = IndexedParameterAttribute()
+        my_par = MyParameter(smirks='[*:1]', a=1)
+        assert my_par.a == 1
+
+    def test_same_length_indexed_attrs(self):
+        """ParameterType raises TypeError if indexed attributes of different lengths are given."""
+        class MyParameter(ParameterType):
+            a = IndexedParameterAttribute()
+            b = IndexedParameterAttribute()
+        with pytest.raises(TypeError, match="indexed attributes have different lengths"):
+            MyParameter(smirks='[*:1]', a1=1, a2=2, a3=3, b1=1, b2=2)
+
+    def test_error_single_value_plus_index(self):
+        """ParameterType raises an error if an indexed attribute is specified with and without index."""
+        class MyParameter(ParameterType):
+            a = IndexedParameterAttribute()
+        with pytest.raises(TypeError, match="'a' has been specified with and without index"):
+            MyParameter(smirks='[*:1]', a=1, a1=2)
```

Comment:
I've implemented it this way only because this is the current behavior and I didn't want to break the API in this PR, but I agree that passing a list is more "pythonic" than passing `a1`, `a2`, which makes sense only for serialization. In fact, I think we could even consider taking only `a` as a list in the normal constructor, and leave the `a1`, `a2`, ... syntax for a `from_dict()` constructor instead for serialization. It seems to me a more standard way of dealing with arguments in Python.

> `a` as a list, we should write a test for it as well.

Good thinking! I'll add it.

Reply:
Although, now that I think about it, it might be a little weird for `BondType.k`, which is currently an indexed attribute.

Reply:
> In fact, I think we could even consider taking only a as a list in the normal constructor, and leave the a1, a2, ... syntax for a from_dict() constructor instead for serialization. It seems to me a more standard way of dealing with arguments in Python.

Agreed. I would be happy with an implementation which, at least for now, only accepts a list (or  iterable) here.